### PR TITLE
[Runtimes] Add support for preemption mode part 2

### PR DESCRIPTION
**Pull request by Tankilevitch**
https://jira.iguazeng.com/browse/ML-1942

Part 1 - https://github.com/mlrun/mlrun/pull/1844

* Add support support for all runtimes except  for spark operator ( part 3 is on the way )
* Add preemption tests for kubejob & nuclio runtimes
* Fix from_dict
* Add from_dict(to_dict()) when mocking kubejob to simulate call to the api

### CODE REVIEW

File: mlrun/runtimes/pod.py [hunk 1]
```diff
@@ -535,6 +518,9 @@ def enrich_function_preemption_spec(self):
                     )
                 )
             )
+            # remove preemptible nodes constrain
+            self._prune_node_selector(mlconf.get_preemptible_node_selector())
```
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/runtimes/pod.py:522]: @Hedingber we are already pruning the node selector in allow mode(as in nuclio) so I thought if we are doing that in allow it makes even more sense to do the same in prevent

[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/runtimes/pod.py:521]: Talked offline - The scenario. of having the preemptible labels in the node selector can potentially happen on upgrade from a version that did not support the preemption mode yet, therefore required the users to specify selectors manually
Therefore I think it indeed does make sense to clean it

File: tests/api/runtimes/test_nuclio.py [hunk 2]
```diff
@@ -253,53 +256,47 @@ def _assert_nuclio_v3io_mount(self, local_path="", remote_path="", cred_only=Fal
...
-                    kube_resource_spec._transform_attribute_to_k8s_class_instance(
-                        "affinity", deploy_spec["affinity"]
-                    ),
-                    expected_affinity,
-                    ignore_order=True,
+                mlrun.runtimes.pod.transform_attribute_to_k8s_class_instance(
+                    "affinity", deploy_spec.get("affinity")
                 )
-                == {}
+                == affinity
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply tests/api/runtimes/test_nuclio.py]: talked offline, the affinity class in a deepdiff with `ignore_order` print those errors
```
ERROR [deepdiff.diff] Can not produce a hash for root._node_affinity._required_during_scheduling_ignored_during_execution._node_selector_terms.Not counting this object.
 not readable
ERROR [deepdiff.diff] Can not produce a hash for iterable root._node_affinity._required_during_scheduling_ignored_during_execution._node_selector_terms. not readable
```
So probably better to use `==` for this class

File: tests/api/runtimes/base.py [hunk 3]
```diff
@@ -311,12 +327,19 @@ def _mock_vault_functionality(self):
...
+        self._execute_run(runtime, **kwargs)
+
     def _execute_run(self, runtime, **kwargs):
         # Reset the mock, so that when checking is create_pod was called, no leftovers are there (in case running
         # multiple runs in the same test)
         get_k8s().v1api.create_namespaced_pod.reset_mock()
         get_k8s().v1api.list_namespaced_pod.reset_mock()
 
+        # simulating sending to API - serialization through dict
+        runtime = runtime.from_dict(runtime.to_dict())
```
Comment (Hedingber) [inline tests/api/runtimes/base.py:342]: Not needed now that it's in `execute_function` no ?

File: tests/api/runtimes/base.py [hunk 4]
```diff
@@ -687,3 +710,444 @@ def assert_run_without_specifying_resources(self):
...
+        runtime.with_preemption_mode(mlrun.api.schemas.PreemptionModes.allow.value)
+        self.execute_function(runtime)
+        self.assert_node_selection()
+
+        logger.info("allow -> prevent, expecting preemptible anti-affinity")
+        runtime.with_preemption_mode(mlrun.api.schemas.PreemptionModes.prevent.value)
+        self.execute_function(runtime)
+        self.assert_node_selection(affinity=self._generate_preemptible_anti_affinity())
+
+        logger.info(
```
Comment (Hedingber) [inline tests/api/runtimes/base.py:975]: Same I would separate from here

File: tests/api/runtimes/test_nuclio.py [hunk 5]
```diff
@@ -52,8 +54,6 @@ def custom_setup(self):
         os.environ["V3IO_USERNAME"] = self.v3io_user = "test-user"
 
     def _serialize_and_deploy_nuclio_function(self, function):
-        # simulating sending to API - serialization through dict
-        function = function.from_dict(function.to_dict())
         deploy_nuclio_function(function)
```
Comment (Hedingber) [inline tests/api/runtimes/test_nuclio.py:57]: What left of `_serialize_and_deploy_nuclio_function` is exactly the same as `_execute_run` so I guess remove it, and use `_execute_run` everywhere

File: tests/api/runtimes/base.py [hunk 6]
```diff
@@ -687,3 +718,558 @@ def assert_run_without_specifying_resources(self):
...
+        runtime.with_preemption_mode(mlrun.api.schemas.PreemptionModes.allow.value)
+        self.execute_function(runtime)
+        self.assert_node_selection()
+
+        logger.info("allow -> prevent, expecting preemptible anti-affinity")
+        runtime.with_preemption_mode(mlrun.api.schemas.PreemptionModes.prevent.value)
+        self.execute_function(runtime)
+        self.assert_node_selection(affinity=self._generate_preemptible_anti_affinity())
+
+    def assert_run_preemption_mode_with_preemptible_node_selector_without_tolerations_with_extra_settings(
```
Comment (Hedingber) [inline tests/api/runtimes/base.py:1046]:
```suggestion
    def assert_run_preemption_mode_with_preemptible_node_selector_without_preemptible_tolerations_with_extra_settings(
```

**Review by Hedingber — APPROVED** [review]

**Pull request merged**